File: Pulse in WR90.m
close
clear

% Constants
f0 = 10e9;      % Center frequency (10 GHz)
t0 = 10e-9;     % Pulse duration (10 ns)
z = 10e-2;      % Propagation distance (10 cm)
c = 3e8;        % Speed of light

% Time and frequency vectors
Fs = 10 * f0;   % Sampling frequency
dt = 1 / Fs;    % Time step
t = -5 * t0 : dt : 5 * t0 - dt;
N = length(t);
df = Fs / N;
f = -Fs/2 : df : Fs/2 - df;

% Create Gaussian pulse
pulse = exp(-t.^2 / t0^2);

% Modulate the pulse
modulated_pulse = pulse .* exp(1i * 2 * pi * f0 * t);

% Propagate the pulse
beta = sqrt(1 - (2 * pi * f / c).^2);
modulated_pulse_z = fftshift(ifft(fftshift(modulated_pulse)) .* exp(-1i * beta * z));

% Plot the pulse shape at z = 10 cm
figure;
plot(t, abs(modulated_pulse_z).^2);
xlabel('Time (s)');
ylabel('Power');
title('Pulse Shape at z = 10 cm');

% Calculate the time it takes for the pulse shape to change significantly
group_delay = -diff(unwrap(angle(modulated_pulse_z))) / (2 * pi * df);
significant_change_time = max(group_delay);

fprintf('The pulse shape changes significantly after %.2f seconds.\n', significant_change_time);

%%
fc = 6.56e9;
Fs=5e10;
t=-50e-9:1/Fs:50e-9;
%%
%create pulse
pulse=zeros(1,length(t));
for i=2250:1:2750
    pulse(1,i)=1;
end
plot(t,pulse)
ylim([-1,2])
xlabel('time')
title('10 ns pulse')
%%      
%modulate the pulse
w0=10e9*2*pi;
mpulse=exp(1i*w0*t).*pulse;
%%
%take fft of modulated pulse
fftpulse=fft(mpulse);
L=length(pulse);
f = Fs*(0:floor((L))-1)/L;
plot(f,abs(fftpulse))
xlabel('frequency')
title('fft of the modulated pulse')
%%
%cancel frequancies below cutoff
for i=1:length(f)
    if (f(i)<fc)
        fftpulse(i)=0;

    end
end
figure
plot(f,abs(fftpulse))
%%
m=1;
for i=1:length(f)
    if (fftpulse(i)==0)
        m=m+1;
    else
        break
    end
end
f2=f(m:end);
fftpulse2=fftpulse(m:end);
%calculating propogation constants
beta=sqrt(1-(fc./f2).^2);
%shape of pulse in z=10
z=1000;
for i=1:length(fftpulse2)
    fftpulse2(i)=fftpulse2(i).*exp(-1i*z*beta(i));
end

%calculate ifft to find the result signal
fftpulse3=zeros(1,L);
fftpulse3(m:end)=fftpulse2;

figure
plot(f,abs(fftpulse3))

%%
%calculate ifft to find the result signal
disppulse=ifft(fftpulse3);
disppulse2=disppulse.*exp(-1i*w0*t);
plot(t,abs(real(disppulse2)))

File: Pulse in WR90.py
import numpy as np
import matplotlib.pyplot as plt

fc = 6.56e9
Fs = 5e10
t = np.arange(-50e-9, 50e-9, 1/Fs)

# Create pulse
pulse = np.zeros(len(t))
pulse[2250:2750] = 1

plt.plot(t, pulse)
plt.ylim([-1, 2])
plt.xlabel('Time')
plt.ylabel('Amplitude')
plt.title('10 ns Pulse')

# Modulate the pulse
w0 = 10e9 * 2 * np.pi
mpulse = np.exp(1j * w0 * t) * pulse

# Take FFT of modulated pulse
fftpulse = np.fft.fft(mpulse)
L = len(pulse)
f = Fs * np.arange(0, np.floor(L)) / L

plt.figure()
plt.plot(f, np.abs(fftpulse))
plt.xlabel('Frequency')
plt.ylabel('Magnitude')
plt.title('FFT of the Modulated Pulse')

# Cancel frequencies below cutoff
fftpulse[~(f >= fc)] = 0

plt.figure()
plt.plot(f, np.abs(fftpulse))
plt.xlabel('Frequency')
plt.ylabel('Magnitude')
plt.title('FFT after Cutoff')

m = np.where(fftpulse != 0)[0][0]
f2 = f[m:]
fftpulse2 = fftpulse[m:]

# Calculate propagation constants
beta = np.sqrt(1 - (fc / f2) ** 2)

# Shape of pulse at z=10
z = 1000
fftpulse2 *= np.exp(-1j * z * beta)

# Calculate IFFT to find the result signal
fftpulse3 = np.zeros(L, dtype=complex)
fftpulse3[m:] = fftpulse2

plt.figure()
plt.plot(f, np.abs(fftpulse3))
plt.xlabel('Frequency')
plt.ylabel('Magnitude')
plt.title('Modified FFT')

# Calculate IFFT to find the result signal
disppulse = np.fft.ifft(fftpulse3)
disppulse2 = disppulse * np.exp(-1j * w0 * t)

plt.figure()
plt.plot(t, np.abs(np.real(disppulse2)))
plt.xlabel('Time')
plt.ylabel('Amplitude')
plt.title('Resulting Signal')

# Set same x-axis limits for figures 2, 3, and 4 as figure 5
fig2, fig3, fig4 = plt.gcf(), plt.gcf(), plt.gcf()
axes2, axes3, axes4 = fig2.get_axes(), fig3.get_axes(), fig4.get_axes()
axes2[0].set_xlim([-5e-8, 5e-8])
axes3[0].set_xlim([-5e-8, 5e-8])
axes4[0].set_xlim([-5e-8, 5e-8])

plt.show()
